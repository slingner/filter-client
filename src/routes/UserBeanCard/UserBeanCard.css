.Bean{
  list-style-type: none;
  font-family: 'Montserrat', sans-serif;
  color: white;
  background-color: rgb(130, 114, 93);
  border: 1px solid grey;
  -webkit-box-shadow: 14px 17px 23px -14px rgba(145,132,116,1);
  -moz-box-shadow: 14px 17px 23px -14px rgba(145,132,116,1);
  box-shadow: 14px 17px 23px -14px rgba(145,132,116,1);
  border-radius: 5px;
  margin: 30px;
  padding: 15px;
  max-width: 300px;
  min-height: 300px;
}

.inputAboutBean {
  border-radius: 5px;
  padding: 5px;
  border-color: rgb(130, 114, 93);
  min-width: 200px;
}

.reviews {
  list-style-type: none;
  text-align: left;
}

.delete-review-button {
  margin-left: 5px;
  padding: 5px;
  background-color: rgb(88, 113, 88);
  color: white;
  border: 1px solid rgb(88, 113, 88);
  font-family: 'Montserrat', sans-serif;
  font-size: 13px;
}

.review-submit-button{
  font-family: 'Montserrat', sans-serif;
  font-size: 16px;
  border-radius: 5px;
}

.delete-bean-card-button {
  font-family: 'Montserrat', sans-serif;
  font-size: 16px;
  border-radius: 5px;
}

