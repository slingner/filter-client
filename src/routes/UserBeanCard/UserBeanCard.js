import React from 'react'
import './UserBeanCard.css'
import config from '../../config';
import TokenService from '../../services/token-service';
import BeansListContext from '../../contexts/BeansListContext';
import FilterApiService from '../../services/filter-api-service';
import Review from '../../components/Review/Review';

export default class BeanPage extends React.Component {

  state = {
    reviews: []
  }

  static contextType = BeansListContext;

  componentDidMount() {
    this.fetchBeanReviews();
  }

  fetchBeanReviews = () => {
    FilterApiService.getBeanReviews(this.props.id)
      .then((reviews) => {
        this.setState({ reviews });
      });
  }

  addReview = review => {
    let items = Array.isArray(review) ? review : review;

    items = items.filter((item) => {
      const hasReview = this.state.reviews.some((review) => {
        return review.id === item.id
      })
      return !hasReview
    });

    this.setState({
      reviews: this.state.reviews.concat(items)
    });
  }

  handleSubmit = (e) => {
    e.preventDefault()
    const { text } = e.target
    FilterApiService.postReview(text.value, this.props.id)
      .then((res)=> {
        this.addReview(res)
      })
      .catch(err => {
        console.error(err);
      })
    e.target.text.value = ''
  }
// fetch delete call to delete specific bean card based on bean.id
  deleteBean = (id) =>{
    const newBeans = this.props.userBeans.filter(beans =>
    beans.id !== id
    )
    this.setState({
      beans: newBeans,
    })
    return fetch(`${config.API_ENDPOINT}/userbean/${id}`, {
      method: 'DELETE',
      headers: {
        'content-type': 'application/json',
        'authorization': `bearer ${TokenService.getAuthToken()}`
      },
    })
    .then(res => {
      if (!res.ok) {
        return res.json().then(error => { 
          throw error
        })
      }
      this.context.fetchBeanByUser();
    })
    .catch(error => {
      console.error(error)
    })
  }

  // fetch delete call to delete specific bean card based on bean id
  deleteReview = (id) =>{
    const newReviews = this.context.reviews.filter(reviews =>
    reviews.id !== id
    )
    this.context.addReview(newReviews)
    return fetch(`${config.API_ENDPOINT}/reviews/${id}`, {
      method: 'DELETE',
      headers: {
        'content-type': 'application/json',
        'authorization': `bearer ${TokenService.getAuthToken()}`
      },
    })
    .then(res => {
      if (!res.ok) {
        return res.json().then(error => { 
          throw error
        })
      }
      this.fetchBeanReviews();
    })
    .catch(error => {
      console.error(error)
    })
  }

  render() {
 
    return (
          <div className='Bean'>
            <h3 className='Bean_name'>
             {this.props.bean_name}
            </h3>
            <p className='Bean_origin'>
             <em>Origin:</em><br /> {this.props.bean_origin}
            </p>
            <p className='Bean_masl'>
            <em>Meters Above Sea Level:</em><br /> {this.props.bean_masl}
            </p>
            <p className='Bean_grower'>
            <em>Grower:</em><br /> {this.props.bean_grower}
            </p>
            <p className='Bean_process'>
            <em>Process:</em><br /> {this.props.bean_process}
            </p>
            <p className='flavor_notes'>
            <em>Flavor Notes:</em><br /> {this.props.flavor_notes}
            </p>
            <form  className='reviewSubmit' onSubmit={this.handleSubmit}>
              <textarea 
                aria-label='Comment about this bean...'
                name='text'
                id={this.props.id}
                cols='25'
                rows='3'
                placeholder='Comment about this bean..'
                className="inputAboutBean">
              </textarea>
              <button className='review-submit-button' type='submit'>Submit Bean Note</button>
            </form>
            <ul className='reviews'>
            <em>Notes On Bean:</em><br />  {this.state.reviews.map((review) => (<Review key={`review-${review.id}`} deleteReview={this.deleteReview} reviewText={review.text} reviewId={review.id} />))}
            </ul>
            <button className='delete-bean-card-button' onClick={() => this.deleteBean(this.props.id)}>Delete Bean</button>
          </div>
        )}
}

// BeanPage.propTypes = {
//   bean_name: PropTypes.string.isRequired,
// }





