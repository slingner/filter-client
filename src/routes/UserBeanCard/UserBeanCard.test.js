import React from 'react';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import UserBeanCard from './UserBeanCard'

describe(`UserBeanCard component`, () => {
  const props = {
    id: '1',
    bean_name: 'testname',
    bean_origin: 'testbeanorigin',
    bean_masl: 'testbeanbeanmasl',
    bean_grower: 'testbeangrower',
    bean_process: 'testbeanprocess',
    flavor_notes: 'testflavornotes',
  }

  it('renders a .UserBeanCard by default', () => {
    const wrapper = shallow(<UserBeanCard />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('renders the UserBeanCard given props', () => {
    const wrapper = shallow(<UserBeanCard {...props} />)
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})