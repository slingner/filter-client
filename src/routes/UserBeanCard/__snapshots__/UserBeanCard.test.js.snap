// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserBeanCard component renders a .UserBeanCard by default 1`] = `
<div
  className="Bean"
>
  <h3
    className="Bean_name"
  />
  <p
    className="Bean_origin"
  >
    <em>
      Origin:
    </em>
    <br />
     
  </p>
  <p
    className="Bean_masl"
  >
    <em>
      Meters Above Sea Level:
    </em>
    <br />
     
  </p>
  <p
    className="Bean_grower"
  >
    <em>
      Grower:
    </em>
    <br />
     
  </p>
  <p
    className="Bean_process"
  >
    <em>
      Process:
    </em>
    <br />
     
  </p>
  <p
    className="flavor_notes"
  >
    <em>
      Flavor Notes:
    </em>
    <br />
     
  </p>
  <form
    className="reviewSubmit"
    onSubmit={[Function]}
  >
    <textarea
      aria-label="Comment about this bean..."
      className="inputAboutBean"
      cols="25"
      name="text"
      placeholder="Comment about this bean.."
      rows="3"
    />
    <button
      className="review-submit-button"
      type="submit"
    >
      Submit Bean Note
    </button>
  </form>
  <ul
    className="reviews"
  >
    <em>
      Notes On Bean:
    </em>
    <br />
      
  </ul>
  <button
    className="delete-bean-card-button"
    onClick={[Function]}
  >
    Delete Bean
  </button>
</div>
`;

exports[`UserBeanCard component renders the UserBeanCard given props 1`] = `
<div
  className="Bean"
>
  <h3
    className="Bean_name"
  >
    testname
  </h3>
  <p
    className="Bean_origin"
  >
    <em>
      Origin:
    </em>
    <br />
     
    testbeanorigin
  </p>
  <p
    className="Bean_masl"
  >
    <em>
      Meters Above Sea Level:
    </em>
    <br />
     
    testbeanbeanmasl
  </p>
  <p
    className="Bean_grower"
  >
    <em>
      Grower:
    </em>
    <br />
     
    testbeangrower
  </p>
  <p
    className="Bean_process"
  >
    <em>
      Process:
    </em>
    <br />
     
    testbeanprocess
  </p>
  <p
    className="flavor_notes"
  >
    <em>
      Flavor Notes:
    </em>
    <br />
     
    testflavornotes
  </p>
  <form
    className="reviewSubmit"
    onSubmit={[Function]}
  >
    <textarea
      aria-label="Comment about this bean..."
      className="inputAboutBean"
      cols="25"
      id="1"
      name="text"
      placeholder="Comment about this bean.."
      rows="3"
    />
    <button
      className="review-submit-button"
      type="submit"
    >
      Submit Bean Note
    </button>
  </form>
  <ul
    className="reviews"
  >
    <em>
      Notes On Bean:
    </em>
    <br />
      
  </ul>
  <button
    className="delete-bean-card-button"
    onClick={[Function]}
  >
    Delete Bean
  </button>
</div>
`;
